### PR TITLE
feat(hierarchical): support sample_weight

### LINKED ISSUE #620: `HierarchicalPredictor` and `HierarchicalTransformer`
# Description

As noted in #616 and discussed on private channels, `GroupedPredictor` class has its shortcomings.

I am opening this issue to keep track on how we could expand from what we learned.
Here a list of features that `HierarchicalPredictor` and `HierarchicalTransformer` should have in my opinion:

1. Given an input group, all levels should be fitted, including a global model
    - Example: `groups = ["a", "b"]`, groups to fit are: `["global"], ["global", "a"], ["global", "a", "b"]`
    - This implies that now the order _always_ matters, not only when shrinkage is not None.
    - This is quite opinionated, but also guarantees that 3.1 (see later) always works.
2. Shrinkage available for **both** regression and classification tasks (tbd how it could be extended to outlier detection)
3. Fallback strategy to be either `"next"` or `"raise"`, meaning:
    - `"next"`: if a group value is not found at prediction/transformation time, it fallbacks to the first available group in the hierarchy.
    - `"raise"`: if a group value is not found at prediction/transformation time, an error is raised.
    - ~Disclaimer: better value instead of "next" can be found 😂~ what about `"parent"`?
4. Allows for extra params in fit (e.g. `sample_weight`), which is currently not possible.

A first draft of the implementation was developed in #618, however trying to keep the current behavior of `GroupedPredictor` as well as expanding its functionalities was going to become a headache both to develop and maintain. This is why an implementation from scratch would make sense.

## Edit

A few considerations regarding the hypothetical `HierarchicalTransformer`:

- Shrinkage doesn't really make sense, or at least not for all transformers. In fact the current `GroupedTransformer` doesn't support it.
- Should the group columns be returned as they are from the `.transform(X)` operation?
- Should the `.transform(X)` output maintain the input type? Hence if pandas dataframe maintain the index  as well?

### PR DESCRIPTION
**Pull request by AhmedThahir**
Confirmed with @FBruzzesi on the direction of the PR. This discussion took take place in #620

# Description

Supporting sample_weight for HierarchicalPredictor, HierarchicalRegressor, HierarchicalClassifier.

Fixes #620

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the style guidelines (ruff)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (also to the readme.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added tests to check whether the new feature adheres to the sklearn convention
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by AhmedThahir** [comment]
@FBruzzesi, this is the best I could do, have not been able to understand how to proceed. If you could take some time to assist, it would be great.

```
=========================================================================== short test summary info ===========================================================================
FAILED tests/test_meta/test_hierarchical_predictor.py::test_sklearn_compatible_estimator[HierarchicalRegressor(estimator=LinearRegression(),groups=0)-check_sample_weights_not_an_array] - ValueError: DataFrame constructor not properly called!
FAILED tests/test_meta/test_hierarchical_predictor.py::test_sklearn_compatible_estimator[HierarchicalRegressor(estimator=LinearRegression(),groups=0)-check_sample_weight_equivalence_on_dense_data] - ZeroDivisionError: Weights sum to zero, can't be normalized
FAILED tests/test_meta/test_hierarchical_predictor.py::test_sklearn_compatible_estimator[HierarchicalRegressor(estimator=LinearRegression(),groups=0)-check_sample_weight_equivalence_on_sparse_data] - ValueError: Estimator does not work on sparse matrices
================================================================= 3 failed, 144 passed, 10 skipped in 11.49s ==================================================================
```

**Comment by FBruzzesi** [comment]
Thanks for the contribution @AhmedThahir 🚀

> @FBruzzesi, this is the best I could do, have not been able to understand how to proceed. If you could take some time to assist, it would be great.

I will take a look later today or tomorrow 😇

File: sklego/meta/hierarchical_predictor.py [hunk 1]
```diff
@@ -220,13 +222,17 @@ def __init__(
         self.n_jobs = n_jobs
         self.check_X = check_X
         self.shrinkage_kwargs = shrinkage_kwargs
+        try:
+            self.estimator_supports_sample_weight = "sample_weight" in inspect.signature(self.estimator.fit).parameters
+        except Exception:
+            self.estimator_supports_sample_weight = False
```
Comment (FBruzzesi) [inline sklego/meta/hierarchical_predictor.py:228]: For a scikit-learn compatible estimator you cannot do any operation in `__init__` method. You can assign such attribute only during fit, and it must be (semi) private, with a trailing `_`

File: sklego/meta/hierarchical_predictor.py [hunk 2]
```diff
@@ -281,12 +289,24 @@ def fit(self, X, y=None):
         if self.n_features_in_ < 1:
             msg = "Found 0 features, while a minimum of 1 if required."
             raise ValueError(msg)
+        
+        self.has_sw_ = sample_weight is not None
+        
+        if self.has_sw_ and not self.estimator_supports_sample_weight:
+            msg = f"Estimator does not support sample_weight."
+            raise ValueError(msg)
+        sample_weight = _check_sample_weight(sample_weight, X, None, ensure_non_negative=True)
```
Comment (FBruzzesi) [inline sklego/meta/hierarchical_predictor.py:298]: Here you probably need to cast X to native since scikit-learn does not (yet) work with narwhals objects, also `dtype` argument has already None as default:

```suggestion
        sample_weight = _check_sample_weight(sample_weight, X.to_native(), ensure_non_negative=True)
```

File: sklego/meta/hierarchical_predictor.py [hunk 3]
```diff
@@ -378,10 +399,15 @@ def _predict_estimators(self, X, method_name):
...
     def _fit_single_estimator(self, grp_frame):
         """Shortcut to fit an estimator on a single group"""
-        _X = nw.to_native(grp_frame.drop([*self.groups_, self._TARGET_NAME]))
+        _X = nw.to_native(grp_frame.drop([*self.groups_, self._TARGET_NAME, self._SAMPLE_WEIGHT_NAME]))
         _y = nw.to_native(grp_frame[self._TARGET_NAME])
-
-        return clone(self.estimator).fit(_X, _y)
+        
+        args = [_X, _y]
+        if self.estimator_supports_sample_weight and self.has_sw_:
```
Comment (FBruzzesi) [inline sklego/meta/hierarchical_predictor.py:406]: `and self.has_sw_` can probably be skipped - passing all ones should behave the same as not passing any sample weight (since the estimator supports them)

**Review by FBruzzesi — COMMENTED** [review]
Hey @AhmedThahir thanks again for the contribution, this is off a great start! I think we are already quite close!

I left a few suggestions in the code. Additionally to those fixes, could you add a few test cases?

**Comment by AhmedThahir** [comment]
Alright!

Will work on the:
1. suggested code changes
2. additional test cases

Will get back to you once done tonight/tomorrow.

Best Regards
Ahmed Thahir
LinkedIn <https://www.linkedin.com/in/AhmedThahir> | YouTube
<https://www.youtube.com/channel/UCZRDn0ZVbEYEHKVmMFBI6zQ>

**Comment by AhmedThahir** [comment]
Still failing these tests when I run `pytest tests/test_meta/test_hierarchical_predictor.py`

```
sklego/meta/_grouped_utils.py:64: ValueError
=========================================================================== short test summary info ===========================================================================
FAILED tests/test_meta/test_hierarchical_predictor.py::test_sklearn_compatible_estimator[HierarchicalRegressor(estimator=LinearRegression(),groups=0)-check_sample_weights_not_an_array] - ValueError: DataFrame constructor not properly called!
FAILED tests/test_meta/test_hierarchical_predictor.py::test_sklearn_compatible_estimator[HierarchicalRegressor(estimator=LinearRegression(),groups=0)-check_sample_weight_equivalence_on_dense_data] - ZeroDivisionError: Weights sum to zero, can't be normalized
FAILED tests/test_meta/test_hierarchical_predictor.py::test_sklearn_compatible_estimator[HierarchicalRegressor(estimator=LinearRegression(),groups=0)-check_sample_weight_equivalence_on_sparse_data] - ValueError: Estimator does not work on sparse matrices
```

**Comment by AhmedThahir** [comment]
Also, @FBruzzesi , don't the default sklearn sample weight tests (which I have failed above) take care of the tests? What other tests would be required in our case?

**Comment by FBruzzesi** [comment]
Hey @AhmedThahir thanks for pushing the changes 

> Still failing these tests when I run `pytest tests/test_meta/test_hierarchical_predictor.py`

I am trying to debug the tests locally. From what I can see:

- `check_sample_weights_not_an_array`  (ValueError: DataFrame constructor not properly called!): raises in line 277 because X is not an array type. I think it's ok to raise since some "structure" in the input is required anyway for Hierarchical predictors, however I would raise a more informative error message.

- `check_sample_weight_equivalence_on_dense_data` (ZeroDivisionError: Weights sum to zero, can't be normalized) - the issue is with the group column being a random continuous value from the automatic sklearn check, hence not really a group-like 

One similar issue I can think of is, what should happen if one `sample_weight` is non-zero, but one group has all zeros in the array subset? I don't have an answer for that, but I think that would be ok to raise.

@koaning considering we are already skipping 9 sklearn compatible estimator check for Hierarchical, I would consider removing that entirely and come up with some ad-hoc tests. What do you think?

**Comment by AhmedThahir** [comment]
Any action required from my side ?

**Comment by AhmedThahir** [comment]
Any action required from my side ?

**Comment by FBruzzesi** [comment]
> Any action required from my side ?

Hey @AhmedThahir, not really! If you are interested you can think about how we can test most of the hierchical functionalities and edge cases eventually

**Comment by koaning** [comment]
> @koaning considering we are already skipping 9 sklearn compatible estimator check for Hierarchical, I would consider removing that entirely and come up with some ad-hoc tests. What do you think?

Yeah, agree. Optimise for ease of maintainance here :)

**Comment by AhmedThahir** [comment]
> > Any action required from my side ?
> 
> Hey @AhmedThahir, not really! If you are interested you can think about how we can test most of the hierchical functionalities and edge cases eventually

I'm not sure if I'm quite sure how to proceed here. the inbuilt check sample weights seems to take care of all cases that I can think of.

**Comment by FBruzzesi** [comment]
> I'm not sure if I'm quite sure how to proceed here. the inbuilt check sample weights seems to take care of all cases that I can think of.

Hey @AhmedThahir, what I meant is bigger scoped actually. For now you might:

- Add the failing tests to the list of those to skip (`check_sample_weights_not_an_array`, `check_sample_weight_equivalence_on_dense_data`, `check_sample_weight_equivalence_on_sparse_data`)
- Introduce new dedicated tests to check sample weight behave as expected in the dataframe/array cases
